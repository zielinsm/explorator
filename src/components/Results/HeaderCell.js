import React from "react";
import { Table } from "semantic-ui-react";

import "./HeaderCell.css";

const HeaderCell = props => {
  const { label, key, value } = props.cell;
  const { cellClick } = props;

  const onClick = function onCellClick(e) {
    cellClick(e, value);
  };

  return (
    <Table.HeaderCell onClick={onClick} className="HeaderCell" key={key}>
      {label}
    </Table.HeaderCell>
  );
};

export default HeaderCell;
