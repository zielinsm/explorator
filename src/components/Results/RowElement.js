import React from "react";
import { Label, Table } from "semantic-ui-react";

import truncate from "../../helpers/truncate";
import "./RowElement.css";

const RowElement = props => {
  const data = props.row;
  const rowClick = props.rowClick;

  const onClick = function onRowClick(e) {
    const { fingerprint } = data;
    rowClick(e, fingerprint);
  };

  const renderLabel = function renderRelayLabel(status) {
    if (status) {
      return <Label color="green">Running</Label>;
    } else {
      return <Label color="red">Offline</Label>;
    }
  };

  const { name, fingerprint, address, running } = data;
  const truncatedFingerprint = truncate(fingerprint, 15);

  return (
    <Table.Row onClick={onClick} className="RowElement">
      <Table.Cell>{name}</Table.Cell>
      <Table.Cell>{address[0]}</Table.Cell>
      <Table.Cell>{truncatedFingerprint}</Table.Cell>
      <Table.Cell>{renderLabel(running)}</Table.Cell>
    </Table.Row>
  );
};

export default RowElement;
