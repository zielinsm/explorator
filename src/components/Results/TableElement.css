.TableElement__pagination {
    margin-top: 0.5em !important;
    margin-bottom: 0.25em !important;
}
