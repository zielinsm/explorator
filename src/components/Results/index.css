.Results__Header {
    margin-top: 1rem !important;
}

.Results__no-results {
    padding-top: 1em !important;
    padding-bottom: 1em !important;
}
