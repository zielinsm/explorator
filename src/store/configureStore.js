import { applyMiddleware, createStore } from "redux";
import thunk from "redux-thunk";

import rootReducer from "../reducers";

function configureStore(preloadedState) {
  const store = createStore(
    rootReducer,
    preloadedState,
    applyMiddleware(thunk)
  );
  return store;
}

export default configureStore;
