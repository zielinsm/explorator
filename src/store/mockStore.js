import configureMockStore from "redux-mock-store";
import thunk from "redux-thunk";

const middlewares = [thunk];
const mockStore = configureMockStore(middlewares);

const mockSearch = [
  {
    name: "mockNode",
    fingerprint: "MOCKNODE",
    address: "000.000.000.000",
    running: true
  },
  {
    name: "mockRelay",
    fingerprint: "MOCKRELAY",
    address: "000.000.000.000",
    running: true
  }
];

const mockNode = {
  name: "Name",
  id: "FINGERPRINT",
  or_addr: ["000.000.000.000"],
  dir_addr: "111.111.111.111",
  up: true,
  details: ["Mock", "Running"],
  country: "Mock",
  as: {
    number: "AS194",
    name: "Mock Networks"
  },
  weight: 10000,
  host: "mock-relay.net",
  bandwidth: {
    rate: 10000,
    burst: 10000,
    observed: 10000,
    advertised: 10000
  },
  policy: ["reject *:*"],
  policy_summary_v4: { reject: ["1-65535"] },
  policy_summary_v6: { reject: ["1-65535"] },
  family_alleged: ["ALLEGED"],
  family_effective: ["EFFECTIVE"],
  info: "Contact",
  tor: "Tor 0.3.2.10 on Mock"
};

const store = mockStore({
  searchReducer: {
    isLoading: false,
    data: mockSearch
  },
  nodeReducer: {
    isLoading: false,
    data: mockNode
  }
});

export default store;
