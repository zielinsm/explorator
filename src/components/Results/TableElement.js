import React, { Component } from "react";
import arraySort from "array-sort";
import PropTypes from "prop-types";
import { Grid, Pagination, Table } from "semantic-ui-react";

import HeaderCell from "./HeaderCell";
import RowElement from "./RowElement";
import "./TableElement.css";

class TableElement extends Component {
  constructor(props) {
    super(props);
    this.data = props.results;
    this.perPage = props.perPage || 10;
    this.headerRow = props.headerRow;
    this.totalPages = Math.ceil(this.data.length / this.perPage);

    this.onSortChange = this.handleSortChange.bind(this);
    this.onRowClick = this.handleRowClick.bind(this);

    this.state = {
      data: this.data,
      activePage: 1,
      sort: "name",
      sortAscending: true
    };
  }

  pagination(data, activePage, perPage) {
    const finalElement = activePage * perPage;
    return data.slice(finalElement - perPage, finalElement);
  }

  sortData(data, column, sortAscending) {
    if (sortAscending) {
      return arraySort(data, column);
    } else {
      return arraySort(data, column, { reverse: true });
    }
  }

  handlePaginationChange = (e, { activePage }) => this.setState({ activePage });

  handleRowClick(e, fingerprint) {
    this.context.router.history.push("/node/" + fingerprint);
  }

  handleSortChange = (e, column) => {
    if (this.state.sort === column) {
      const opposite = !this.state.sortAscending;
      this.setState({ sortAscending: opposite });
    } else {
      this.setState({ sort: column });
    }
  };

  renderHeaderCell(data, index, handler) {
    return <HeaderCell cell={data} cellClick={handler} key={index} />;
  }

  renderBodyRow(data, handler) {
    return <RowElement key={data.fingerprint} row={data} rowClick={handler} />;
  }

  renderPagination(current, total, handler) {
    if (total > 1) {
      return (
        <Grid centered className="TableElement__pagination">
          <Pagination
            size="mini"
            activePage={current}
            totalPages={total}
            onPageChange={handler}
          />
        </Grid>
      );
    } else return null;
  }

  render() {
    const sorted = this.sortData(
      this.data,
      this.state.sort,
      this.state.sortAscending
    );
    const page = this.pagination(sorted, this.state.activePage, this.perPage);

    return (
      <div className="TableElement">
        <Table selectable>
          <Table.Header>
            <Table.Row>
              {this.headerRow.map((cell, index) =>
                this.renderHeaderCell(cell, index, this.onSortChange)
              )}
            </Table.Row>
          </Table.Header>
          <Table.Body>
            {page.map(row => this.renderBodyRow(row, this.onRowClick))}
          </Table.Body>
        </Table>
        {this.renderPagination(
          this.state.activePage,
          this.totalPages,
          this.handlePaginationChange
        )}
      </div>
    );
  }

  static contextTypes = {
    router: PropTypes.object
  };
}

export default TableElement;
