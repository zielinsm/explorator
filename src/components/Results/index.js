import React, { PureComponent } from "react";
import { connect } from "react-redux";
import { Link } from "react-router-dom";
import { Breadcrumb, Header, Loader, Segment } from "semantic-ui-react";

import { fetchSearch } from "../../actions/";
import TableElement from "./TableElement";
import "./index.css";

function Navigation() {
  return (
    <Breadcrumb className="Node__Breadcrumbs">
      <Breadcrumb.Section>
        <Link to="/">Search</Link>
      </Breadcrumb.Section>
      <Breadcrumb.Divider />
      <Breadcrumb.Section active>Results</Breadcrumb.Section>
    </Breadcrumb>
  );
}

const headerRow = [
  { label: "Name", value: "name" },
  { label: "IP", value: "ip" },
  { label: "Fingerprint", value: "fingerprint" },
  { label: "Status", value: "running" }
];

class Results extends PureComponent {
  constructor(props) {
    super(props);
    this.routeParameter = props.match.params.id;
  }

  componentDidMount() {
    const { fetchSearch } = this.props;
    if (this.routeParameter) {
      fetchSearch(this.routeParameter);
    }
  }

  renderDefaultContent(data) {
    if (data.length > 0) {
      return <TableElement results={data} headerRow={headerRow} />;
    } else {
      return (
        <Header as="h5" className="Results__no-results" textAlign="center">
          No results, try searching again...
        </Header>
      );
    }
  }

  render() {
    const { searchReducer } = this.props;
    const { data, isLoading } = searchReducer;

    if (!isLoading) {
      return (
        <div>
          <Navigation />
          <Header as="h3" attached="top" className="Results__Header">
            Showing results for "{this.routeParameter}"
          </Header>
          <Segment attached>{this.renderDefaultContent(data)}</Segment>
        </div>
      );
    } else {
      return <Loader active>Fetching data...</Loader>;
    }
  }
}

const mapState = function mapStateToProps(state) {
  const { searchReducer } = state;
  return { searchReducer };
};

export default connect(
  mapState,
  { fetchSearch }
)(Results);
