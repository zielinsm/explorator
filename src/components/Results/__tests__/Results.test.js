import React from "react";
import renderer from "react-test-renderer";
import { Provider } from "react-redux";
import { MemoryRouter } from "react-router";
import store from "../../../store/mockStore";

import Results from "../index";

describe("Results container:", () => {
  let wrapper;
  const component = (
    <Provider store={store}>
      <MemoryRouter>
        <Results match={{ params: { id: "mockTerm" } }} />
      </MemoryRouter>
    </Provider>
  );

  it("snapshot matches", () => {
    const tree = renderer.create(component).toJSON();
    expect(tree).toMatchSnapshot();
  });
});
