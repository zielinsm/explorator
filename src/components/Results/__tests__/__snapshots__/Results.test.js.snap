// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Results container: snapshot matches 1`] = `
<div>
  <div
    className="ui breadcrumb Node__Breadcrumbs"
  >
    <div
      className="section"
      href={undefined}
      onClick={[Function]}
    >
      <a
        href="/"
        onClick={[Function]}
      >
        Search
      </a>
    </div>
    <div
      className="divider"
    >
      /
    </div>
    <div
      className="active section"
      href={undefined}
      onClick={[Function]}
    >
      Results
    </div>
  </div>
  <h3
    className="ui top attached header Results__Header"
  >
    Showing results for "
    mockTerm
    "
  </h3>
  <div
    className="ui attached segment"
  >
    <div
      className="TableElement"
    >
      <table
        className="ui selectable table"
      >
        <thead
          className=""
        >
          <tr
            className=""
          >
            <th
              className="HeaderCell"
              onClick={[Function]}
            >
              Name
            </th>
            <th
              className="HeaderCell"
              onClick={[Function]}
            >
              IP
            </th>
            <th
              className="HeaderCell"
              onClick={[Function]}
            >
              Fingerprint
            </th>
            <th
              className="HeaderCell"
              onClick={[Function]}
            >
              Status
            </th>
          </tr>
        </thead>
        <tbody
          className=""
        >
          <tr
            className="RowElement"
            onClick={[Function]}
          >
            <td
              className=""
            >
              mockNode
            </td>
            <td
              className=""
            >
              0
            </td>
            <td
              className=""
            >
              MOCKNODE
            </td>
            <td
              className=""
            >
              <div
                className="ui green label"
                onClick={[Function]}
              >
                Running
              </div>
            </td>
          </tr>
          <tr
            className="RowElement"
            onClick={[Function]}
          >
            <td
              className=""
            >
              mockRelay
            </td>
            <td
              className=""
            >
              0
            </td>
            <td
              className=""
            >
              MOCKRELAY
            </td>
            <td
              className=""
            >
              <div
                className="ui green label"
                onClick={[Function]}
              >
                Running
              </div>
            </td>
          </tr>
        </tbody>
      </table>
    </div>
  </div>
</div>
`;
